### ISSUE TITLE
Android app is killed in low memory scenario (very large texture cache)

### ISSUE THREAD

**Issue by aardvarkk** [cocos2d/cocos2d-x#17345]
- cocos2d-x version: 3.14.1
- devices test on: Motorola Moto G 2014 (CyanogenMod), Sony Xperia
- developing environments
   - NDK version: r13b
   - Xcode version: N/A
   - VS version: N/A
   - browser type and version: N/A

Steps to Reproduce:

1. Across several Scenes, load and display large textures
2. Over time, the texture cache fills up and eventually the Android app is killed due to low memory

Our app has several large images (basically full screen textures) we were using across different scenes. We are experiencing crashes on lower end devices that seem directly related to memory usage. An examination of `adb log` just shows the message: `process has died`. There's no signal sent to the debbuger -- no actual crash occurs. The OS just kills the app because of excess memory usage.

On the Java side, I added custom callbacks to `onTrimMemory` which make a call through the JNI to Cocos to call `removeUnusedTextures` in the texture cache. Although it seems to help a small amount, the problem is not fixed. The app will still be killed randomly.

What I think would help tremendously would be some kind of limit on how large the texture cache could get. Instead of continually loading textures in, it would only keep textures in the cache up to a memory limit and after that point it would continually release the least-recently-used texture until it was able to load the requested texture.

I will look into implementing such a feature, but wanted to create the issue in case other people are having a similar problem with their app.

**Comment by aardvarkk** [comment]
Is there code anywhere to even *try* to clear out the texture cache in some kind of automated manner, or will it always get larger as the game continues to run and new textures are loaded?

**Comment by minggo** [comment]
> Is there code anywhere to even try to clear out the texture cache in some kind of automated manner

Did you mean VolatileTextureMgr?

**Comment by minggo** [comment]
The app should take charge of the memory. Apps is better to add a callback when there is any memory warning message sent by app. And please use forum for such discussion instead of issue system.

**Comment by aardvarkk** [comment]
I haven't seen any examples of `VolatileTextureMgr` -- is there a way to limit the amount of memory textures can occupy with it? I don't see any public functions that would imply that functionality.

I understand your belief that the app should handle the memory, but the problem is that it's basically impossible to do this on Android (from what I've seen). The reason I filed this issue is that *even if I call `removeAllTextures` when Android calls `onTrimMemory`, my app is still killed by the OS!*

So I'm doing what you're saying: I get a memory warning, I clear the texture cache -- but the process is *still* killed. I'm guessing it might take a frame or two for the app to clear out the memory? But regardless, the approach doesn't work.

I ended up adding a function `setMemoryLimit(int limit)` to `TextureCache`. This sets a soft limit on the amount of memory the `TextureCache` can occupy, as opposed to its current approach which is just to get larger and larger without bound (as far as I can tell). Whenever `addImage` is called, it checks whether the texture cache is over the limit, and if it is, it removes any unused textures before adding the new image. It seems to work quite well for me.
